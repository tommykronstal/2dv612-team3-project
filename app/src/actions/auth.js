import { SET_LOGGED_IN, LOGOUT } from './types'
import { getPayloadFromJwtToken } from '../lib/jwt'

export const setUserToLoggedIn = response => ({ 
  type: SET_LOGGED_IN, 
  token: response.token, 
  payload: getPayloadFromJwtToken(response.token) 
})

export const logout = () => ({ type: LOGOUT })
