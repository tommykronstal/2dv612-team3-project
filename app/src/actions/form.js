import {CLEAR_FORM, SUBMIT_FORM, UPDATE_FIELD} from './types'

export const submitForm = (form, endpoint, action, tokenRequire = true, role = '') => ({
  type: SUBMIT_FORM,
  form,
  endpoint,
  action,
  tokenRequire,
  role
})

export const clearForm = form => ({
  type: CLEAR_FORM,
  form
})

export const updateField = (form, field, value) => ({
  type: UPDATE_FIELD,
  form,
  field,
  value
})
