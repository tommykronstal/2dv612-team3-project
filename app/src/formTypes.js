export const LOGIN = 'LOGIN'
export const ADD_COMPANY = 'ADD_COMPANY'
export const ADD_COMPANY_REP = 'ADD_COMPANY_REP'
export const REGISTER_CONSUMER = 'REGISTER_CONSUMER'
export const ADD_PRODUCT = 'ADD_PRODUCT'
export const UPDATE_PRODUCT = 'UPDATE_PRODUCT'
export const ADD_CATEGORY = 'ADD_CATEGORY'