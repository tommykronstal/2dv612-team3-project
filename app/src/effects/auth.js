import {takeEvery} from 'redux-saga/effects'
import {SET_LOGGED_IN, LOGOUT} from '../actions/types'
import {storeJwtToken, resetJwtToken} from '../lib/jwt'

export function *watchAuthenticationActions () {
  yield takeEvery(SET_LOGGED_IN, storeJWT)
  yield takeEvery(LOGOUT, resetJWT)
}

export function *storeJWT ({ token }) {
  yield storeJwtToken(token)
}

export function *resetJWT () {
  yield resetJwtToken()
}
