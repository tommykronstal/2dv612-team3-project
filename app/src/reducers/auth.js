import {SET_LOGGED_IN, LOGOUT} from '../actions/types'
import { getJwtToken, getPayloadFromJwtToken } from '../lib/jwt'

const token = getJwtToken()
const data = getPayloadFromJwtToken(token)

const defaultState = {
  isAuthenticated: !!token,
  jwt: token,
  email: data.email,
  firstName: data.firstName,
  role: data.role,
  userId: data.userId,
  companyId: data.companyId
}

export default (state = defaultState, action) => {
  switch (action.type) {
  case SET_LOGGED_IN:
    return {
      ...state,
      isAuthenticated: true,
      jwt: action.token,
      email: action.payload.email,
      firstName: action.payload.firstName,
      role: action.payload.role,
      userId: action.payload.userId,
      companyId: action.payload.companyId
    }
  case LOGOUT:
    return {
      isAuthenticated: false,
      jwt: null,
      email: null,
      firstName: null,
      role: null,
      userId: null,
      companyId: null
    }
  default:
    return state
  }
}
