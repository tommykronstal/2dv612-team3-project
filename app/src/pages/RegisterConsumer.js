import React from 'react'
import Title from '../components/common/Title'
import Input from '../components/common/Input'
import Content from '../components/common/Content'
import Button from '../components/common/Button'
import {connect} from 'react-redux'
import {REGISTER_CONSUMER} from '../formTypes'
import {submitForm, updateField} from '../actions/form'
import {setUserToLoggedIn} from '../actions/auth'
import {Redirect} from 'react-router-dom'

export const RegisterConsumer = ({
  updateField,
  tryRegister,
  loading: {isLoading},
  form: {email, firstname, lastname, password},
  auth: {isAuthenticated},
}) =>
  isAuthenticated ? (
    <Redirect to='/' />
  ) : (
    <Content>
      <form
        onSubmit={e => {
          e.preventDefault()
          tryRegister()
        }}
      >
        <Title>Register</Title>
        <Input
          type='email'
          label='E-mail'
          name='email'
          value={email}
          onChange={updateField}
        />
        <Input
          type='text'
          label='Firstname'
          name='firstname'
          value={firstname}
          onChange={updateField}
        />
        <Input
          type='text'
          label='Lastname'
          name='lastname'
          value={lastname}
          onChange={updateField}
        />
        <Input
          type='password'
          label='Pasword'
          name='password'
          value={password}
          onChange={updateField}
        />
        <Button primary loading={isLoading}>
          Register
        </Button>
      </form>
    </Content>
  )

export const mapStateToProps = ({auth, loading, form}) => ({
  auth,
  loading,
  form: form[REGISTER_CONSUMER] || {},
}) 

export const mapDispatchToProps = dispatch => ({
  tryRegister: () =>
    dispatch(
      submitForm(
        REGISTER_CONSUMER,
        '/api/user/register',
        setUserToLoggedIn,
        false
      ),
    ),
  updateField: ({target: {value, name}}) =>
    dispatch(updateField(REGISTER_CONSUMER, name, value)),
}) 

export default connect(mapStateToProps, mapDispatchToProps)(RegisterConsumer)
