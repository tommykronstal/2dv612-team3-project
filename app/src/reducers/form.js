import { CLEAR_FORM, UPDATE_FIELD } from '../actions/types'
import { REGISTER_CONSUMER } from '../formTypes'
import USER_TYPES from '../userTypes'

const initialState = {
  [REGISTER_CONSUMER]: {
    role: USER_TYPES.USER
  }
}

export default function status (state = initialState, action = {}) {
  switch (action.type) {
  case CLEAR_FORM:
    return {
      ...state,
      [action.form]: initialState[action.form] || {}
    }
  case UPDATE_FIELD:
    return {
      ...state,
      [action.form]: {
        ...state[action.form],
        [action.field]: action.value
      }
    }
  default:
    return state
  }
}
