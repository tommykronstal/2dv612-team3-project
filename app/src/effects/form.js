import {put, takeEvery, select, call} from 'redux-saga/effects'
import {SUBMIT_FORM, TOGGLE_LOADING, SET_STATUS} from '../actions/types'
import {clearForm} from '../actions/form'
import {post} from '../lib/http'
import {UPDATE_PRODUCT} from '../formTypes'

export function* watchFormActions() {
  yield takeEvery(SUBMIT_FORM, formRequest)
}

export function* formRequest({endpoint, form, action, tokenRequired}) {
  yield put({type: TOGGLE_LOADING})


  const {token, payload} = yield select(state => ({
    token: state.auth.jwt,
    payload: {
      ...state.form[form],
    },
  }))

  const tokenHeader = {...(token && tokenRequired && {Authorization: token})}

  if (form === UPDATE_PRODUCT) {
    yield sendProductFile(endpoint, token, payload, tokenHeader, action, form)
    return
  }

  console.log(payload)

  const response = yield call(post, endpoint, {
    headers: {
      ...tokenHeader,
    },
    body: JSON.stringify(payload)
  })

  console.log('response', response)

  yield handleResponse(response, form, action)
}

function* handleResponse(response, form, action) {
  if (response.status === 200 || response.status === 201) {
    if (action) {
      yield put(action(response))
    }
    yield put(clearForm(form))
  } else {
    if (response.message)
      yield put({type: SET_STATUS, warning: true, message: response.message})
  }

  yield put({type: TOGGLE_LOADING})
}


/**
 * Used to send data as a form.
 * Need to send it like this since original function to submit forms
 * assumes that it's JSON. (Maybe needs some tweaking in the future...)
 */
function* sendProductFile(endpoint, token, payload, tokenHeader = {}, action, form) {
  const formData = new FormData()

  console.log(payload)

  for( let key in payload) {
    formData.append(key, payload[key])
  }

  const response = yield call(post, endpoint, {
    headers: {
      ...tokenHeader
    },
    isJsonPayload: false,
    body: formData
  })

  yield handleResponse(response, form, action)
}