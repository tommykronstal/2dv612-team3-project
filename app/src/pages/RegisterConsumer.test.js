import React from 'react'
import renderer from 'react-test-renderer'

import { RegisterConsumer, mapStateToProps, mapDispatchToProps } from './RegisterConsumer'
import { setUserToLoggedIn } from '../actions/auth'
import { REGISTER_CONSUMER } from '../formTypes'
import { submitForm, updateField } from '../actions/form'

describe('RegisterConsumer', () => {

  describe('component', () => {
    it('renders correctly, initial', () => {
      const props = {
        tryRegister: jest.fn(),
        updateField: jest.fn(),
        form: {},
        loading: { isLoading: false },
        auth: { isAuthenticated: false }
      }
  
      const tree = renderer.create(<RegisterConsumer {...props} />).toJSON()
  
      expect(tree).toMatchSnapshot()
    })
  
    it('renders correctly, with data', () => {
      const props = {
        tryRegister: jest.fn(),
        updateField: jest.fn(),
        form: {
          firstName: 'Kalle',
          lastName: 'Kula', 
          email: 'woo@woo.com', 
          password: 'aaa' 
        },
        loading: { isLoading: true },
        auth: { isAuthenticated: false }
      }
  
      const tree = renderer.create(<RegisterConsumer {...props} />).toJSON()
  
      expect(tree).toMatchSnapshot()
    })
  })

  describe('state', () => {
    it('should have access to auth, register form and loading state', () => {
      const state = {
        auth: { token: 'wooo' },
        loading: { isLoading: false },
        form: { [REGISTER_CONSUMER]: {username: 'woooo'}}
      }
    
      expect(mapStateToProps(state)).toEqual({ 
        auth: state.auth, 
        loading: state.loading, 
        form: state.form[REGISTER_CONSUMER] 
      })
    })
  })

  describe('actions', () => {
    it('should be able to dispatch submit form actions', () => {
      const dispatch = jest.fn()
    
      const props = mapDispatchToProps(dispatch)
      props.tryRegister()
    
      expect(dispatch).toBeCalledWith(submitForm(REGISTER_CONSUMER, '/api/user/register', setUserToLoggedIn, false))
    })

    it('should be able to dispatch update field actions', () => {
      const dispatch = jest.fn()
      const event = {
        target: {
          name: 'user',
          value: 'kalle'
        }
      }
    
      const props = mapDispatchToProps(dispatch)
      props.updateField(event)
    
      expect(dispatch).toBeCalledWith(updateField(REGISTER_CONSUMER, event.target.name, event.target.value))
    })
  })
})
