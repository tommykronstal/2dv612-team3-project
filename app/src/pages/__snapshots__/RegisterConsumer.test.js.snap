// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterConsumer component renders correctly, initial 1`] = `
<div
  className="sc-EHOje eCCCFh"
>
  <form
    className="sc-bZQynM XFITM"
    onSubmit={[Function]}
  >
    <section
      className="sc-ifAKCX fkzLlq"
    >
      <h1
        className="sc-bdVaJa jbqxFm"
      >
        Register
      </h1>
      <div
        className="sc-bxivhb gZceFu"
      >
        <input
          className="sc-bwzfXH cdlyNt"
          name="email"
          onChange={[Function]}
          type="email"
          value=""
        />
        <label
          className="sc-htpNat dJdhXc"
          value=""
        >
          E-mail
        </label>
      </div>
      <div
        className="sc-bxivhb gZceFu"
      >
        <input
          className="sc-bwzfXH cdlyNt"
          name="firstname"
          onChange={[Function]}
          type="text"
          value=""
        />
        <label
          className="sc-htpNat dJdhXc"
          value=""
        >
          Firstname
        </label>
      </div>
      <div
        className="sc-bxivhb gZceFu"
      >
        <input
          className="sc-bwzfXH cdlyNt"
          name="lastname"
          onChange={[Function]}
          type="text"
          value=""
        />
        <label
          className="sc-htpNat dJdhXc"
          value=""
        >
          Lastname
        </label>
      </div>
      <div
        className="sc-bxivhb gZceFu"
      >
        <input
          className="sc-bwzfXH cdlyNt"
          name="password"
          onChange={[Function]}
          type="password"
          value=""
        />
        <label
          className="sc-htpNat dJdhXc"
          value=""
        >
          Pasword
        </label>
      </div>
      <button
        className="sc-htoDjs gyJotO"
        disabled={false}
        onClick={[Function]}
      >
        Register
      </button>
    </section>
  </form>
</div>
`;

exports[`RegisterConsumer component renders correctly, with data 1`] = `
<div
  className="sc-EHOje eCCCFh"
>
  <form
    className="sc-bZQynM XFITM"
    onSubmit={[Function]}
  >
    <section
      className="sc-ifAKCX fkzLlq"
    >
      <h1
        className="sc-bdVaJa jbqxFm"
      >
        Register
      </h1>
      <div
        className="sc-bxivhb gZceFu"
      >
        <input
          className="sc-bwzfXH cdlyNt"
          name="email"
          onChange={[Function]}
          type="email"
          value="woo@woo.com"
        />
        <label
          className="sc-htpNat iOTirm"
          value="woo@woo.com"
        >
          E-mail
        </label>
      </div>
      <div
        className="sc-bxivhb gZceFu"
      >
        <input
          className="sc-bwzfXH cdlyNt"
          name="firstname"
          onChange={[Function]}
          type="text"
          value=""
        />
        <label
          className="sc-htpNat dJdhXc"
          value=""
        >
          Firstname
        </label>
      </div>
      <div
        className="sc-bxivhb gZceFu"
      >
        <input
          className="sc-bwzfXH cdlyNt"
          name="lastname"
          onChange={[Function]}
          type="text"
          value=""
        />
        <label
          className="sc-htpNat dJdhXc"
          value=""
        >
          Lastname
        </label>
      </div>
      <div
        className="sc-bxivhb gZceFu"
      >
        <input
          className="sc-bwzfXH cdlyNt"
          name="password"
          onChange={[Function]}
          type="password"
          value="aaa"
        />
        <label
          className="sc-htpNat iOTirm"
          value="aaa"
        >
          Pasword
        </label>
      </div>
      <button
        className="sc-htoDjs kYfdvH"
        disabled={true}
        onClick={[Function]}
      >
        <div>
          <div
            className="sc-gzVnrw iUTliq"
            size={0.5}
          />
          <div
            className="sc-gzVnrw bpLhQz"
            size={0.5}
          />
          <div
            className="sc-gzVnrw dOBXYC"
            size={0.5}
          />
        </div>
      </button>
    </section>
  </form>
</div>
`;
